### PR TITLE
CDAP-8632 Added retry strategy for the Metrics table creation.

### PR DESCRIPTION
**Pull request by sagarkapare**
When `OperationalStatsService` is started, it loads `CDAPLoad` extension which requires to access the metrics table. However if the DatasetService is not up yet, we see following exceptions in the log

```java
2017-02-20 18:00:39,081 - WARN  [OperationalStatsService:c.c.c.m.s.DefaultMetricDatasetFactory@128] - Cannot access or create table metrics.v2.table.ts.1, will retry in 1 sec.
2017-02-20 18:00:41,082 - WARN  [OperationalStatsService:c.c.c.m.s.DefaultMetricDatasetFactory@128] - Cannot access or create table metrics.v2.table.ts.1, will retry in 1 sec.
2017-02-20 18:00:43,082 - WARN  [OperationalStatsService:c.c.c.m.s.DefaultMetricDatasetFactory@128] - Cannot access or create table metrics.v2.table.ts.1, will retry 
```

In this PR added fix delay retry strategy for the creation of the metrics table, since `OperationalStatsService` is a system service.

### CODE REVIEW

File: cdap-watchdog/src/main/java/co/cask/cdap/metrics/store/DefaultMetricDatasetFactory.java [hunk 1]
```diff
@@ -114,30 +123,23 @@ public MetricsConsumerMetaTable createConsumerMeta() {
...
+        @Override
+        public MetricsTable call() throws Exception {
+          return DatasetsUtil.getOrCreateDataset(dsFramework, metricsDatasetInstanceId, MetricsTable.class.getName(),
+                                                 props, null, null);
         }
-      } catch (IOException e) {
-        LOG.error("Exception while creating table {}.", tableName, e);
-        throw Throwables.propagate(e);
-      }
+      }, RetryStrategies.fixDelay(1, TimeUnit.SECONDS), RETRYABLE_PREDICATE);
```
Comment (chtyim) [inline cdap-watchdog/src/main/java/co/cask/cdap/metrics/store/DefaultMetricDatasetFactory.java]: I am a bit concerning in doing retry for this. Because this could mean any call to the REST API for getting metrics will hang until the retry give up if DatasetService is not available. A better behavior for the REST API is to retry for much smaller amount of times (say at most 2 seconds) and then response with Service Unavailable if still failing.

Comment (chtyim) [reply cdap-watchdog/src/main/java/co/cask/cdap/metrics/store/DefaultMetricDatasetFactory.java]: For failing in getting dataset, I think it's better to have caller to decide what to do. E.g. for the operation stat case, it can simply just catch the service unavailable exception (assuming we propagate properly) and ignore it (without logging) as it'll get retried in the next interaction anyway.

The same code path is used by the metrics system. For the metrics processor, it might want to keep retrying until it succeed as we don't want losing any metrics. It can be done as part of the polling logic in the metrics processor when it is polling from TMS (simply just don't advance the offset if the writing was not successful). This means retry shouldn't be done in this method.

For the metrics handler, as mentioned above, retry is actually not needed (or only a very short retry).

Comment (sagarkapare) [reply cdap-watchdog/src/main/java/co/cask/cdap/metrics/store/DefaultMetricDatasetFactory.java]: @chtyim Addressed comments.
1. Not retrying in the `DefaultMetricDatasetFactory` instead relying on the caller.
2. In `OperationalStatsService`, catching the `ServiceUnavailableException` and not logging. This will be tried in the next iteration anyway.
3. We were already retrying by sleeping for 1 second in the `KafkaMetricsProcessorService` and `MessagingMetricsProcessorService`. Catching the `ServiceUnavailableException` there so that we do not log unnecessary.

**Comment by sagarkapare** [comment]
@chtyim Thanks for the review. Addressed comments. Can you please take one more look.

File: cdap-app-fabric/src/main/java/co/cask/cdap/operations/OperationalStatsService.java [hunk 2]
```diff
@@ -126,6 +127,11 @@ private void collectOperationalStats() throws InterruptedException {
       } catch (InterruptedException e) {
         throw e;
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/operations/OperationalStatsService.java]: It's not part of this PR, but catch and rethrow like this is useless.

File: cdap-app-fabric/src/main/java/co/cask/cdap/operations/OperationalStatsService.java [hunk 3]
```diff
@@ -126,6 +127,11 @@ private void collectOperationalStats() throws InterruptedException {
...
       } catch (Throwable t) {
+        if (t instanceof ServiceUnavailableException) {
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/operations/OperationalStatsService.java]: You probably need to get the root cause

File: cdap-watchdog/src/main/java/co/cask/cdap/metrics/store/DefaultMetricDatasetFactory.java [hunk 4]
```diff
@@ -59,16 +57,10 @@
   private final DatasetFramework dsFramework;
 
   @Inject
-  public DefaultMetricDatasetFactory(final CConfiguration cConf, final DatasetFramework dsFramework) {
-    this(dsFramework, cConf);
```
Comment (chtyim) [inline cdap-watchdog/src/main/java/co/cask/cdap/metrics/store/DefaultMetricDatasetFactory.java:63]: This is a good one. How can that happen.

File: cdap-watchdog/src/main/java/co/cask/cdap/metrics/store/DefaultMetricDatasetFactory.java [hunk 5]
```diff
@@ -115,29 +107,15 @@ public MetricsConsumerMetaTable createConsumerMeta() {
...
-      } catch (IOException e) {
-        LOG.error("Exception while creating table {}.", tableName, e);
-        throw Throwables.propagate(e);
-      }
+    MetricsTable table = null;
+    try {
+      table = DatasetsUtil.getOrCreateDataset(dsFramework, metricsDatasetInstanceId, MetricsTable.class.getName(),
+                                             props, null, null);
+    } catch (Exception e) {
+      Throwables.propagate(e);
```
Comment (chtyim) [inline cdap-watchdog/src/main/java/co/cask/cdap/metrics/store/DefaultMetricDatasetFactory.java]: The caller to this method already try catch and wrap. Better just have this method `throws Exception`.

Comment (chtyim) [reply cdap-watchdog/src/main/java/co/cask/cdap/metrics/store/DefaultMetricDatasetFactory.java]: Should also get rid of the `LOG.error` in the `createConsumerMeta` method and change the `LOG.info` to debug.

File: cdap-app-fabric/src/main/java/co/cask/cdap/operations/OperationalStatsService.java (hunk 2 above)
Comment (sagarkapare) [reply cdap-app-fabric/src/main/java/co/cask/cdap/operations/OperationalStatsService.java]: Removed this catch block and updated following catch block.
```java
catch (Throwable t) {
   Throwables.propagateIfInstanceOf(t, InterruptedException.class);
   ...
}
```

File: cdap-app-fabric/src/main/java/co/cask/cdap/operations/OperationalStatsService.java (hunk 3 above)
Comment (sagarkapare) [reply cdap-app-fabric/src/main/java/co/cask/cdap/operations/OperationalStatsService.java]: fixed.

File: cdap-watchdog/src/main/java/co/cask/cdap/metrics/store/DefaultMetricDatasetFactory.java (hunk 4 above)
Comment (sagarkapare) [reply cdap-watchdog/src/main/java/co/cask/cdap/metrics/store/DefaultMetricDatasetFactory.java:63]: Previously we were wrapping dsFramework into NamespacedDatasetFramework, which is why we had private constructor, however wrapping got removed in this commit - https://github.com/caskdata/cdap/commit/6e9a6654d40879a14a60313270321cb6b8b9eb8d#diff-7cf678edd77bdb39c8ce63abdb509246L57

So we don't need private constructor anymore.

File: cdap-watchdog/src/main/java/co/cask/cdap/metrics/store/DefaultMetricDatasetFactory.java (hunk 5 above)
Comment (sagarkapare) [reply cdap-watchdog/src/main/java/co/cask/cdap/metrics/store/DefaultMetricDatasetFactory.java]: To avoid duplication of try catch and wrap, doing it in the `getOrCreateMetricsTable` only. Removed wrapping from the `createConsumerMeta`

**Comment by sagarkapare** [comment]
Build is green - https://builds.cask.co/browse/CDAP-RUT653-6

File: cdap-app-fabric/src/main/java/co/cask/cdap/operations/OperationalStatsService.java [hunk 6]
```diff
@@ -123,10 +125,14 @@ private void collectOperationalStats() throws InterruptedException {
       LOG.debug("Collecting stats for service {} of type {}", stats.getServiceName(), stats.getStatType());
       try {
         stats.collect();
-      } catch (InterruptedException e) {
-        throw e;
       } catch (Throwable t) {
+        Throwables.propagateIfInstanceOf(t, InterruptedException.class);
         Throwable rootCause = Throwables.getRootCause(t);
+        if (rootCause instanceof ServiceUnavailableException || rootCause instanceof TException) {
```
Comment (albertshau) [inline cdap-app-fabric/src/main/java/co/cask/cdap/operations/OperationalStatsService.java:131]: TException is from tx client errors?

Comment (albertshau) [reply cdap-app-fabric/src/main/java/co/cask/cdap/operations/OperationalStatsService.java:131]: would be good to add a comment about it

**Review by albertshau — APPROVED** [review]
lgtm

File: cdap-app-fabric/src/main/java/co/cask/cdap/operations/OperationalStatsService.java (hunk 6 above)
Comment (sagarkapare) [reply cdap-app-fabric/src/main/java/co/cask/cdap/operations/OperationalStatsService.java:131]: updated comment on the next line.

**Comment by sagarkapare** [comment]
Build is already green here - https://builds.cask.co/browse/CDAP-RUT653-6. Not running build for the minor comment modification. RAT passes. squashing and merging.
